### PR TITLE
Extended PO parsing logic with validation rules (resolves #75)

### LINKED ISSUE #75: Custom validation rules
Hello :wave:

To start off I would like to thank you for developing and providing the library to the community!

I have a feature request. As a part of automation process of working with PO files I need to impose validation rules while parsing the files, like ensuring there are no duplicate `msgid`s or no entries with the same value for `msgstr`. The library already has a parser and a lexer, and I'd like to be able to reuse them for my particular needs. Two approaches come to my mind:

- the library exposes abovementioned entities as a part of public API,
- the library provides new API that would allow to define custom validation rules.

I'd like to learn you opinion and suggestions. Would you be willing to accept a PR?

### PR DESCRIPTION
**Pull request by vkhytskyi-allegro**
The aim of this PR is to introduce validation rules for PO parsing logic. Those rules check that:

- there is exactly zero or one `msgid_plural` definition per translation entry; a `Multiple msgid_plural error` error gets thrown otherwise.
- there are no duplicate entries with exact `msgid` values; a `Duplicate msgid error` error gets thrown otherwise.
- the number of plural forms matches exactly the number from `nplurals` defined in `Plural-Forms` header for entries that have plural forms; a `Plural forms range error` error gets thrown otherwise.
- the number of `msgstr` matches exacty the one (if `msgid_plural` is not defined) or the number from `nplurals` (if `msgid_plural` is defined); a `Translation string range error` error gets thrown otherwise.

### CODE REVIEW

**Review by szczepanekpawel — APPROVED** [review]

File: test/po-parser-test.js
Comment (KubaZ) [inline test/po-parser-test.js:158]: Why not check the entire error message?

File: test/po-parser-test.js
Comment (KubaZ) [inline test/po-parser-test.js:164]: Why not check the entire error message?

File: test/shared.js
Comment (KubaZ) [inline test/shared.js:183]: Should it return fallback value in this case or rather throw an error? I would opt for the latter

File: test/shared.js
Comment (KubaZ) [inline test/shared.js:190]: Should it return fallback value in this case or rather throw an error? I would opt for the latter

**Comment by smhg** [comment]
@vkhytskyi-allegro Can you please also add usage information to the README? That helps me to understand the changes too.

**Comment by vkhytskyi-allegro** [comment]
@smhg The aim of this PR is to introduce validation rules for PO parsing logic. Those rules check that:

- there is exactly zero or one `msgid_plural` definition per translation entry; a `Multiple msgid_plural error` error gets thrown otherwise.
- there are no duplicate entries with exact `msgid` values; a `Duplicate msgid error` error gets thrown otherwise.
- the number of plural forms matches exactly the number from `nplurals` defined in `Plural-Forms` header for entries that have plural forms; a `Plural forms range error` error gets thrown otherwise.
- the number of `msgstr` matches exacty the one (if `msgid_plural` is not defined) or the number from `nplurals` (if `msgid_plural` is defined); a `Translation string range error` error gets thrown otherwise.

Since the idea of the PR is to enable those rules permanently, I am not sure whether there is anything to add to `README`. 
Please correct me if I am wrong.

File: test/shared.js
Comment (vkhytskyi-allegro) [reply test/shared.js:183]: Yeah, I also have doubts whether `nplurals` parsing function should try parsing it safely and fallback to default value or throw an error. I decided to go with safe variant and leave it for Sam to judge. @smhg what do you think?

**Comment by vkhytskyi-allegro** [comment]
Hey @smhg 👋🏼 Is there a chance you could take a look at my last reply? Thanks!

**Comment by smhg** [comment]
Thank you for following up on this @vkhytskyi-allegro. I did not find the time yet to move this forward.
I will try to add questions when possible. These are mainly meant for me to get insights into this PR. Don't interpret them as any criticism. My responsibility is to make sure the impact these changes have on other (current and future) users are sensible.

File: .eslintrc.js
Comment (smhg) [inline .eslintrc.js:17]: Can you please explain why this was necessary? Do any of the new additions require this?

File: test/shared.js
Comment (smhg) [reply test/shared.js:183]: @KubaZ @vkhytskyi-allegro Is a negative value any different from something like 'foo' (see next test)? Or do did you consider them to be different because of the `parseInt()` return value check (inside `parseNPluralFromHeadersSafely`)?

Comment (smhg) [reply test/shared.js:183]: I'd stick with the fallback approach. Validating headers might be a good thing, but I think we shouldn't start only validating (= throwing an error) one part (`nplurals`) of one header (`Plural-Forms`).

**Comment by smhg** [comment]
I haven't yet looked at the actual `_validateToken` method but shouldn't we put validation behind a flag?
While we could release this as a new major version with it being enabled by default, users will likely still want to turn this off.
@vkhytskyi-allegro @KubaZ @szczepanekpawel Does that sound reasonable?

**Comment by vkhytskyi-allegro** [comment]
> I will try to add questions when possible. These are mainly meant for me to get insights into this PR. Don't interpret them as any criticism. My responsibility is to make sure the impact these changes have on other (current and future) users are sensible.

Sure, I absolutely understand that and open to make changes in the implementation the way you see fit 🙂

File: .eslintrc.js
Comment (vkhytskyi-allegro) [reply .eslintrc.js:17]: The necessity arose because the validation logic has to read keys and values of translation entries from PO files. One of those keys - `msgid_plural` - is written in snake_case. I was considering disabling the rule by adding comments where necessary, but because that occurs in a number of places I decided to go with an update in `.eslintrc.js` file. Should I solve it by adding comments wherever needed?

**Comment by vkhytskyi-allegro** [comment]
@smhg / @KubaZ / @szczepanekpawel the PR is ready for another round of review 😉

Changes since the last time:
- Moved validation logic behind a flag
- Updated README explaining what validation does and how to turn in on
- Added more unit tests
- Reverted ESLint configuration change
- Improved JSDoc

**Review by KubaZ — APPROVED** [review]

File: lib/poparser.js
Comment (smhg) [inline lib/poparser.js:20]: At first sight, I'd prefer to use a method signature like this:
`parse = function (buffer, options) {`
With `options` being an object with optional properties like `defaultCharset` and `validation`. To prevent breaking current usage, i'd say a `typeof options === 'string'` check (meaning `options` is actually`defaultCharset`) would be enough.
In my opinion this facilitates additional options in the future while not needing the 'magic' `unshiftOptionalArgs` provides. But please tell me if I'm missing something with this proposal.

File: lib/poparser.js
Comment (smhg) [inline lib/poparser.js:549]: See above. A method signature like `PoParserTransform (options, streamOptions)` would be better I think. Meaning: the current `options` parameter will become `streamOptions` containing only the options passed specifically for `Transform` below.

Comment (vkhytskyi-allegro) [reply lib/poparser.js:549]: Thank you for your review.

Personally, I am not sure whether having a single `options` argument instead of two dedicated ones would make much difference for a library user. The argument order is still preserved and, hence, is backward compatible. At the same time, I think that from the perspective of project maintainer using control flow directly instead of `unshiftOptionalArgs` would complicate argument parsing quite significant. Here is a list of singatures `stream` function must comply a logic I came up with:

```js
/**
 * Parses a PO stream, emits translation table in object mode
 *
 * @typedef {{ defaultCharset: strubg, validation: boolean }} Options
 * @param defaultCharset Default charset to use
 * @param transformOptions Stream options
 * @param options Optional options with defaultCharset and validation
 * @type {{
*    () => Stream;
*    (defaultCharset: string) => Stream;
*    (defaultCharset: string, transformOptions: import('readable-stream').TransformOptions) => Stream;
*    (transformOptions: import('readable-stream').TransformOptions) => Stream;
*    (options: Options) => Stream;
*    (options: Options, transformOptions: import('readable-stream').TransformOptions) => Stream;
* }}
*/
module.exports.stream = function (arg0, arg1) {
  const [defaultCharset, validation, transformOptions] = typeof arg0 === 'string'
    ? [arg0, undefined, arg1]
    : 'defaultCharset' in arg0
      ? [arg0.defaultCharset, arg0.validation, arg1]
      : [undefined, undefined, arg0];

  return new PoParserTransform(defaultCharset, transformOptions, validation);
};
```

I am not very fond of this approach and I would be curious to learn if there is a better way to solve this. Regardless, if you think the above snippet is the right way to go, please let me know and I'll apply it at once.

Thanks.

Comment (smhg) [reply lib/poparser.js:549]: I might not be skilled enough to have this discussion - my appologies if I'm missing something.
My point of view is: considering plain JS use (not thinking about TS features or JSDoc-based code completion), arguments that can be used in any order (based on their types being different) are confusing. These methods (`parse` and `stream`) will likely get more arguments in the future.
To solve this, an options object is common (as properties are not mandatory and their existence can easily be checked), no?
Or is this an outdated view? Are you aware of other libraries using an `unshiftOptionalArgs` approach (or similar)?

Comment (vkhytskyi-allegro) [reply lib/poparser.js:591]: I see, thanks for sharing your point of view. I made necessary adjustments so please take another look.

File: lib/poparser.js
Comment (vkhytskyi-allegro) [reply lib/poparser.js]: https://github.com/smhg/gettext-parser/pull/76#discussion_r1134071829

**Review by KubaZ — APPROVED** [review]

File: lib/poparser.js
Comment (smhg) [reply lib/poparser.js:591]: Thanks, I think this makes things more simple. I'll write a few (last) comments.

File: lib/poparser.js
Comment (smhg) [inline lib/poparser.js:91]: First of all, I think only 2 signatures should be 'promoted':
* current usage: defaultCharset, transformOptions
* new usage: options, transformOptions
`transformOptions` should never be the first argument.

As the new signature looks way better (thinking about future options), let's write it like that:
```javascript
function stream (options, transformOptions) { 
```
The code then becomes easier to reason about (`arg0` and `arg1` don't have any meaning).

I'll leave a comment below (with `PoParserTransform`) to clarify more.

File: lib/poparser.js
Comment (smhg) [inline lib/poparser.js:94]: Let's use the same pattern here:
```js
PoParserTransform(options, transformOptions);
```
This means `function stream` above does not need anything, just pass on the arguments.

`PoParserTransform` itself would then need to become something like:

```js
function PoParserTransform (options, transformOptions) {
  if (typeof options === 'string') {
    options = { defaultCharset: options };
  }

  const { defaultCharset, validation } = options;
```

I hope this works. Feel free to say otherwise.
The idea is this would eliminated the need for `normalizeStreamArgs` as there is only 1 special use case (the old, `defaultCharset`) and that should not obfuscate the recommended new usage. Although anyone who cares can clearly see it this way (and of course, please keep the remaining type definitions in jsdoc comments - thanks for those!).

File: lib/poparser.js
Comment (vkhytskyi-allegro) [reply lib/poparser.js:91]: > `transformOptions` should never be the first argument.

If I am reading these [lines of code](https://github.com/smhg/gettext-parser/blob/master/lib/poparser.js#L479-L482) correctly, I am afraid having `transformOptions` as the first argument is already permitted. Correct me if I'm wrong but my understanding is that we would like the function signature to stay fully backward compatible. That is the reason for using `arg0` and `arg1`.

I think that if this change is planned to be released with library's major version bump then it would make sense to introduce breaking changes, to forbid having `transformOptions` as the first argument and only support those function signatures:

```js
() => Stream;
(defaultCharset: string) => Stream;
(defaultCharset: string, transformOptions: import('readable-stream').TransformOptions) => Stream;
(options: Options) => Stream;
(options: Options, transformOptions: import('readable-stream').TransformOptions) => Stream;
```

Please let me know if that is the case.

File: lib/poparser.js
Comment (vkhytskyi-allegro) [reply lib/poparser.js:94]: @smhg I hope you don't mind if share my motivation for solving it the way I did it.

It is partially described in https://github.com/smhg/gettext-parser/pull/76#discussion_r1136781289 and, to add to that, I believe this is a good occasion to apply single responsibility principle. I think that parsing incoming arguments should be decoupled from creating an instance of a class. That is why I decided to introduce arguments normalization functions and keep the constructor of `PoParserTransform` a bit thinner. Because those normalization functions are pure I, subjectively, find it to be closer to functional programming style (which, according to our discussion in the relevant [issue](https://github.com/smhg/gettext-parser/issues/75#issuecomment-1344327213), I believe is desirable). Additionally. as a consequence of such decoupling we are able to easily provide fine grained unit test coverage. If we decide to couple those two things back together, I believe testing that would become a bit less convenient.

Given the above, what do you think about it?

**Comment by vkhytskyi-allegro** [comment]
@smhg Sorry for impatience but I'd like to follow up on last comments I left (https://github.com/smhg/gettext-parser/pull/76#discussion_r1136781289 & https://github.com/smhg/gettext-parser/pull/76#discussion_r1136834716) and ask whether you had a chance to take a look at them? Thanks!

**Comment by smhg** [comment]
@vkhytskyi-allegro I will find some time during the next few days to get back to you.
You are obviously right about [these lines](https://github.com/smhg/gettext-parser/blob/master/lib/poparser.js#L479-L482) containing options magic. I didn't know about those (I've adopted this project). I currently prefer to clean that up (major version bump) to enable the behavior I described earlier. Feel free to move ahead or wait for more feedback (if applicable).

File: lib/poparser.js
Comment (vkhytskyi-allegro) [reply lib/poparser.js]: @smhg In accordance with https://github.com/smhg/gettext-parser/pull/76#issuecomment-1478410116 I made necessary [adjustments](https://github.com/smhg/gettext-parser/pull/76/commits/caa544530d6c04cf1e0fddebae0a18f2a88bfa51).

File: lib/poparser.js
Comment (vkhytskyi-allegro) [reply lib/poparser.js:52]: https://github.com/smhg/gettext-parser/pull/76#discussion_r1144566489

**Comment by smhg** [comment]
@vkhytskyi-allegro Thank you for your ongoing work!
I'd like to simplify things further. Since this will be a major release, it makes no sense to still support `defaultCharset` directly (outside of `options.defaultCharset`).
Secondly, I think `Parser` and `PoParserTransform` should handle their options in exactly the same way. Thinking about future options it will just become a longer and longer parameter list otherwise.
Third, the Readme would need to be adapted to reflect the new arguments.
Do you still feel like moving on with this or do you prefer I handle these changes?

**Comment by vkhytskyi-allegro** [comment]
@smhg Thank you for getting back to me. I am open to continue working on the PR till it's ready to be merged. I've just pushed changes that should address your latest comments.

**Comment by smhg** [comment]
Thank you for your contributions @vkhytskyi-allegro! I'll release this soon.

**Comment by KubaZ** [comment]
Hey @smhg 👋 Do You plan to release this change in near future?

**Comment by smhg** [comment]
@KubaZ waiting for another PR, I lost track of this release. I'm sorry.
It's now released as 7.0.0.

**Comment by KubaZ** [comment]
No worries, thanks for quick response and for the release! 🚀

**Pull request merged**
